packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -867,6 +886,235 @@ Please replace "${contractName}" for the correct contract name wherever you are
   }
 }
 
+export class Artifacts implements IArtifacts {
+  private readonly _sourcesInPriorityOrder: [
+    IHardhatArtifactSource,
+    ...ArtifactSource[]
+  ];
```

Comment:
Also: see my other comment about where to put the core Hardhat logic for getting artifacts. If we go with the option of having that logic as part of the `Artifacts` class (or its sub-layers) then this would become just a single variable with the array of extensions.